driver_test.go

```diff
@@ -12,26 +12,250 @@
-		},
+var(
+	project string
+	instance string
+	dbname string
+	dsn string
+)
+
+// connector things 
+type Connector struct {
+	ctx         context.Context	
+	client      *spanner.Client
+	adminClient *adminapi.DatabaseAdminClient
+}
+
+func NewConnector()(*Connector, error){
+
+	ctx := context.Background()
+
+	adminClient, err := adminapi.NewDatabaseAdminClient(
+		ctx,
+		option.WithoutAuthentication(),
+		option.WithEndpoint("0.0.0.0:9010"),
+		option.WithGRPCDialOption(grpc.WithInsecure()))
+	if err != nil {
+			return nil, err
 	}
-	connector, err := driver.OpenConnector("projects/$PROJECT/instances/$INSTANCE/databases/$DATABASE")
+
+	dataClient, err := spanner.NewClient(ctx, dsn)
 	if err != nil {
+			return nil,err
+	}
+
+	curs := &Connector{
+		ctx: ctx,
+		client: dataClient,
+		adminClient: adminClient,
+
+	}
+	return curs,nil
+}
+
+func (c *Connector) Close() {
+	c.client.Close()
+	c.adminClient.Close()
+}
+
+// structs for row data 
+type testaRow struct{
+	A string
+	B string
+	C string
+}
+type typeTestaRow struct {
+	stringt string 
+	intt int 
+	floatt float64
+	boolt bool
+}
+
+func init(){
+
+	// get environment variables
+	instance = os.Getenv("SPANNER_TEST_INSTANCE")
+	project = os.Getenv("SPANNER_TEST_PROJECT")
+	dbname = os.Getenv("SPANNER_TEST_DBNAME")
+
+	// set defaults if none provided 
+	if instance == "" {instance = "test-instance" }
+	if project == "" { project = "test-project" }
+	if dbname == "" { dbname = "gotest" }
+
+	// derive data source name 
+	dsn = "projects/" + project + "/instances/" + instance + "/databases/" + dbname
+}
+
+
+// functions that use the client lib / apis ~ 
+// ******************* //
+
+// Executes DDL statements 
+func executeDdlApi(curs *Connector, ddls []string){
+
+	op, err := curs.adminClient.UpdateDatabaseDdl(curs.ctx, &adminpb.UpdateDatabaseDdlRequest{
+		Database:   dsn,
+		Statements: ddls,
+	})
+	if err != nil {
+		//return nil, err
+		log.Fatal(err)
+	}
+	if err := op.Wait(curs.ctx); err != nil {
+		//return nil, err
 		log.Fatal(err)
 	}
-	db := sql.OpenDB(connector)
-	_ = db // Use db.
 }
+
+// executes DML using the client library 
+func ExecuteDMLClientLib(dml []string){
+
+	// open client
+	var db = "projects/"+project+"/instances/"+instance+"/databases/"+dbname;
+	ctx := context.Background()
+	client, err := spanner.NewClient(ctx, db)
+	if err != nil {
+			log.Fatal(err)
+	}
+	defer client.Close()
+
+	// Put strings into spanner.Statement structure
+	var states []spanner.Statement
+	for _,line := range dml {
+		states = append(states, spanner.NewStatement(line))
+	}
+
+	// execute statements
+	_, err = client.ReadWriteTransaction(ctx, func(ctx context.Context, txn *spanner.ReadWriteTransaction) error {
+		stmts := states
+		rowCounts, err := txn.BatchUpdate(ctx, stmts)
+		if err != nil {
+				return err
+		}
+		fmt.Printf("Executed %d SQL statements using Batch DML.\n", len(rowCounts))
+		return nil
+		})
+	if (err != nil) { log.Fatal(err) }
```

Comment:
In Go, we don't need brackets

```suggestion
	if err != nil { log.Fatal(err) }
```

Reply:
Pretty sure running go fmt took care of this too 🤔